### PR TITLE
Fixing scoping for unification rules

### LINKED ISSUE #455: Bug in unification rule scoping
When scoping the RHS of a unification rule, underscores (and thus implicit arguments) should be substituted by fresh pattern variables.
The rationale is that a rule (including a unification rule) added in a signature must not contain any metavariable.

### PR DESCRIPTION
**Pull request by gabrielhdt**
Fixes #455. Introduce a new scoping mode for unification rules.

### CODE REVIEW

File: src/core/scope.ml [hunk 1]
```diff
@@ -253,6 +256,15 @@ let scope : mode -> sig_state -> env -> p_term -> tbox = fun md ss env t ->
...
         find_qid false false ss env qid
+    | (P_Wild, M_URHS(data)) ->
+      let x =
+        let name = Printf.sprintf "v%i" data.m_urhs_vars_nb in
+        let x = Bindlib.new_var te_mkfree name in
+        data.m_urhs_vars_nb <- data.m_urhs_vars_nb + 1;
+        data.m_urhs_xvars <- (name, x) :: data.m_urhs_xvars;
+        x
+      in
+      _TEnv (Bindlib.box_var x) [||]
```
Comment (fblanqui) [inline src/core/scope.ml:267]: Why is the array argument empty? You should take the environment into account. After the doc: "_ for an unknown term or a term we don’t care about. It is replaced by a fresh metavariable (or a fresh pattern variable in a rule left-hand side) applied to all the variables of the context." For instance, in `\x:N,\Pi y:_,a`, _ is equivalent to `&Y[x]` where Y is fresh.

File: src/core/scope.ml [hunk 2]
```diff
@@ -499,9 +529,9 @@ type pre_rule =
   ; pr_xvars_nb : int
   (** Number of variables that appear in the RHS but not in the LHS. *) }
 
-(** [scope_rule ss r] turns a parser-level rewriting rule [r] into a rewriting
+(** [scope_rule u ss r] turns a parser-level rewriting rule [r] into a rewriting
```
Comment (fblanqui) [inline src/core/scope.ml:532]: u -> ur, like in the code below
explanation of ur missing

File: src/core/scope.ml (hunk 1 above)
Comment (gabrielhdt) [reply src/core/scope.ml]: Indeed it's a mistake.

File: src/core/scope.ml [hunk 3]
```diff
@@ -499,9 +529,10 @@ type pre_rule =
   ; pr_xvars_nb : int
   (** Number of variables that appear in the RHS but not in the LHS. *) }
 
-(** [scope_rule ss r] turns a parser-level rewriting rule [r] into a rewriting
-    rule (and the associated head symbol). *)
-let scope_rule : sig_state -> p_rule -> pre_rule loc = fun ss r ->
+(** [scope_rule ur ss r] turns a parser-level rewriting rule [r], or a
+    unification rule if [ur] is true, into a rewriting rule (and the
```
Comment (fblanqui) [inline src/core/scope.ml:533]: into a pre-rule

**Comment by fblanqui** [comment]
Thanks to add Rehan's example in tests/OK to check the PR.

**Comment by gabrielhdt** [comment]
His example did not check anything, or only visually. The file `unif_hint.lp` still fails because of #466. So there is no systematic way to check that it works. To convince yourself, you can run `dune exec -- lambdapi check tests/OK/unif_hint.lp --verbose 3` with `set flag "print_implicits" on` in the file to see that there is no meta anymore.

**Comment by Rehan-MALAK** [comment]
Just to clarify :
-there was no new example, it was just a shortened version of unif_hint.
-unif_hint needs at least three things (and maybe more) to work when we typecheck meta instanciation : change Kind for the correct type (#465), change scoping (#467) and unification is missing constraints (#466)

we removed #465 and #466 during review of #454 because :
- the way I was adding constraints had to be investigated more (it was a bit artificial for this specific example)
- I had no modifications for rule RHS as it's done in #467

**Comment by fblanqui** [comment]
It's not only visual. You should actually be able to check that it computes correctly. Just add a conversion test between the lhs and the rhs. This test should fail in master.

**Comment by gabrielhdt** [comment]
I'm afraid I can't manage to create such an example. I tried something like
```
symbol T: TYPE
injective symbol El: T → TYPE
symbol iota: T
symbol f {t: T} (x: El t): El t
symbol g {t: T} (x: El t): El t
symbol h {t: T} (x: El t): El t
set unif_rule f $x ≡ g $y ↪ h $x ≡ h $y
symbol x: El iota
rule h {iota} _ ↪ x
constant symbol eq {t: T}: El t → El t → TYPE
symbol refl {t: T} (x: El t): eq x x
set flag "print_implicits" on
assert refl (f x): eq (f x) (g x)
```

but in both cases, I end up with `Cannot solve h ?1 x ≡ h ?3 x.`. So the correction seems tricky to test since occurrences of the bug were likely shadowed by the fact that the type checker couldn't infer an implicit argument?

**Comment by fblanqui** [comment]
Right. Currently we cannot ask to evaluate a term without type checking it first, and unification rules use symbols inaccessible to users and with no type.

**Comment by fblanqui** [comment]
@gabrielhdt could you please update this branch so that we can merge it ?

**Comment by gabrielhdt** [comment]
It is updated, isn't it?

**Comment by gabrielhdt** [comment]
Sorry, it wasn't indeed, it is now.

**Comment by fblanqui** [comment]
Thanks @gabrielhdt

**Pull request merged**